File: algorithm/python/prob1001-1500/prob1490_medium_Clone-N-ary-Tree.py
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children if children is not None else []

class Solution:
    def cloneTree(self, root: "Node") -> "Node":
        if root is None:
            return None
        return self.dfs(root)

    def dfs(self, node: "Node") -> "Node":
        new_node = Node(node.val)
        n_child = len(node.children)
        new_node.children = [None] * n_child
        for i, child in enumerate(node.children):
            new_node.children[i] = self.dfs(child)
        return new_node

import copy

# class Solution:
#     def cloneTree(self, root: 'Node') -> 'Node':
#         return copy.deepcopy(root)

import types

def __deepcopy__(self, memo):
    if self in memo:
        return memo.get(self)
    new_root = Node(copy.deepcopy(self.val, memo)
                   ,copy.deepcopy(self.children, memo))
    memo[self] = new_root
    return new_root

class Solution:
    def cloneTree(self, root: 'Node') -> 'Node':
        if root is None:
            return None
        root.__deepcopy__ = types.MethodType(__deepcopy__, root)
        return copy.deepcopy(root)

File: algorithm/cpp/prob1001-1500/prob1490_medium_Clone-N-ary-Tree.cpp
// prob1490: Clone N-ary Tree

/*
 * https://leetcode-cn.com/problems/clone-n-ary-tree/
 */

#include <vector>

using namespace std;

/*
// Definition for a Node.
*/
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

class Solution {
public:
    Node* cloneTree(Node* root) {
        if(!root) return nullptr;
        return build(root);
    }

private:
    Node* build(Node* node)
    {
        Node *ans = new Node(node -> val);
        int m = (node -> children).size();
        (ans -> children).assign(m, nullptr);
        for(int i = 0; i < m; ++i)
        {
            Node *child = (node -> children)[i];
            if(child)
                ans -> children[i] = build(child);
        }
        return ans;
    }
};
